src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1792,20 +1829,112 @@ private void setItemToCacheFromRefresh(final RefreshItem refreshItem,
+    }
+    
+    /**
+     * @return True if access token redeem with RT is done through FRT token cache entry. 
+     */
+    private boolean tokenRefreshedWithFRT(final AuthenticationRequest authRequest, final RefreshItem refreshItem) {
+        return refreshItem != null && refreshItem.mKey.equalsIgnoreCase(CacheKey.createFamilyRefreshTokenKey(authRequest, 
+                getUserFromAuthenticationRequest(authRequest)));
+    }
+
+    /**
+     * Remove resource specific token cache entry.
+     */
+    private void removeResourceSpecificTokenEntry(final AuthenticationRequest authRequest, final RefreshItem refreshItem) {
         if (mTokenCacheStore != null) {
-            Logger.v(TAG, "Remove refresh item from cache:" + refreshItem.mKey);
-            mTokenCacheStore.removeItem(refreshItem.mKey);
-            // clean up keys related to userid/displayableid for same request
-            mTokenCacheStore.removeItem(refreshItem.mKeyWithUserId);
-            mTokenCacheStore.removeItem(refreshItem.mKeyWithDisplayableId);
+            Logger.v(TAG, "Remove token cache entry with resource in the cache key");
+            
+            final String cacheKeyUsed = CacheKey.createCacheKey(authRequest, getUserFromAuthenticationRequest(authRequest));
+            mTokenCacheStore.removeItem(cacheKeyUsed);
+            
+            final String cacheKeyWithDisplayableId = CacheKey.createCacheKey(authRequest, refreshItem.mUserInfo.getDisplayableId());
+            mTokenCacheStore.removeItem(cacheKeyWithDisplayableId);
+            
+            final String cacheKeyWithUniqueUserId = CacheKey.createCacheKey(authRequest, refreshItem.mUserInfo.getUserId());
+            mTokenCacheStore.removeItem(cacheKeyWithUniqueUserId);
+        }
+    }
+    
+    /**
+     * Remove MRRT specific token cache entry.
+     */
+    private void removeMRRTokenEntry(final AuthenticationRequest authRequest, final RefreshItem refreshItem) {
+        if (mTokenCacheStore != null) {
+            Logger.v(TAG, "Remove MRRT token cache entry.");
+            
+            final String cacheKeyUsed = CacheKey.createMultiResourceRefreshTokenKey(authRequest, getUserFromAuthenticationRequest(authRequest));
+            mTokenCacheStore.removeItem(cacheKeyUsed);
+            
+            final String cacheKeyWithDisplayableId = CacheKey.createMultiResourceRefreshTokenKey(authRequest, refreshItem.mUserInfo.getDisplayableId());
+            mTokenCacheStore.removeItem(cacheKeyWithDisplayableId);
+            
+            final String cacheKeyWithUniqueUserId = CacheKey.createMultiResourceRefreshTokenKey(authRequest, refreshItem.mUserInfo.getUserId());
+            mTokenCacheStore.removeItem(cacheKeyWithUniqueUserId);
+        }
+    }
+    
+    /**
+     * Remove FRT token cache entry.
+     */
+    private void removeFamilyTokenEntry(final AuthenticationRequest authRequest, final RefreshItem refreshItem) {
+        if (mTokenCacheStore != null) {
```

Comment:
is there a chance where mTokenCacheStore could be null when this method is called?